ffi/src/lib.rs

```diff
@@ -654,29 +760,61 @@ pub unsafe extern "C" fn visit_schema(
     fn visit_struct_fields(visitor: &EngineSchemaVisitor, s: &StructType) -> usize {
         let child_list_id = (visitor.make_field_list)(visitor.data, s.fields.len());
         for field in s.fields() {
-            visit_field(visitor, child_list_id, field);
+            visit_schema_item(field.data_type(), field.name(), visitor, child_list_id);
         }
         child_list_id
     }
 
+    fn visit_array_item(visitor: &EngineSchemaVisitor, at: &ArrayType) -> usize {
+        let child_list_id = (visitor.make_field_list)(visitor.data, 1);
+        visit_schema_item(&at.element_type, "array_element", visitor, child_list_id);
+        child_list_id
+    }
+
+    fn visit_map_types(visitor: &EngineSchemaVisitor, mt: &MapType) -> usize {
+        let child_list_id = (visitor.make_field_list)(visitor.data, 2);
+        visit_schema_item(&mt.key_type, "map_key", visitor, child_list_id);
+        visit_schema_item(&mt.value_type, "map_value", visitor, child_list_id);
+        child_list_id
+    }
+
     // Visit a struct field (recursively) and add the result to the list of siblings.
-    fn visit_field(visitor: &EngineSchemaVisitor, sibling_list_id: usize, field: &StructField) {
-        let name: &str = field.name.as_ref();
-        match &field.data_type {
-            DataType::Primitive(PrimitiveType::Integer) => {
-                (visitor.visit_integer)(visitor.data, sibling_list_id, name.into())
-            }
-            DataType::Primitive(PrimitiveType::Long) => {
-                (visitor.visit_long)(visitor.data, sibling_list_id, name.into())
-            }
-            DataType::Primitive(PrimitiveType::String) => {
-                (visitor.visit_string)(visitor.data, sibling_list_id, name.into())
+    fn visit_schema_item(
+        data_type: &DataType,
+        name: &str,
+        visitor: &EngineSchemaVisitor,
+        sibling_list_id: usize,
+    ) {
+        macro_rules! call {
+            ( $visitor_fn:ident $(, $extra_args:expr) *) => {
+                (visitor.$visitor_fn)(visitor.data, sibling_list_id, name.into() $(, $extra_args) *)
+            };
+        }
+        match data_type {
+            DataType::Struct(st) => call!(visit_struct, visit_struct_fields(visitor, st)),
+            DataType::Map(mt) => call!(
+                visit_map,
+                mt.value_contains_null,
+                visit_map_types(visitor, mt)
+            ),
```

Comment:
Out of curiosity, what decides whether to have `=> call!(` vs. `=> {`?
```suggestion
            DataType::Map(mt) => {
                call!(visit_map, mt.value_contains_null, visit_map_types(visitor, mt))
            }
```

Reply:
not entirely sure, but i think for `fmt` the trailing `;` does, at least if there is no macro involved.

Reply:
Yeah, `rustfmt` is a bit odd here.  It'll let me put the map call into `{}`s, but won't let it stay on one line. 

It seems to use `{}`s if the body of the match case needs to go on its own line, so

`&DataType::STRING => call!(visit_string),` is okay because it's all one line, but

```rust
DataType::Array(at) =>
    call!(visit_array, at.contains_null, visit_array_item(visitor, at)),
```

while valid, will be `fmt`'d to:

```rust
DataType::Array(at) => {
    call!(visit_array, at.contains_null, visit_array_item(visitor, at))
}
```

I've put the `map` case into `{}`s for a bit more consistency, but really can't find a formatting that will let it stay on one line, I think because `fmt` thinks it's just a bit too long.